### PR TITLE
Fix action handler typing

### PR DESCRIPTION
**Pull request by altano**
Action handlers can currently have invalid states and actions. There are unit tests to prevent this but unfortunately the `@ts-expect-error` was catching unrelated issues, and these parameters aren't currently checked. I've gone ahead and made the fixes required to type check the state and action here correctly.

The only user-facing change here is that I had to change `ErrorBrand`'s implementation and this will affect user-facing errors (but I hope you will agree they're more readable now).

### CODE REVIEW

File: src/StateMachine.ts [hunk 1]
```diff
@@ -31,7 +31,7 @@ export type Action<Name extends ActionNameType, Payload> = Readonly<Payload> & {
  * and subverting error checking. Yet, the compiler still displays the string at the end of the failed cast indicating what the
  * issue is rather than something puzzling like could not assign to never.
  */
-type ErrorBrand<T extends IndexType> = { [k in T]: void };
+type ErrorBrand<T> = T | never
```
Comment (rickwebiii) [inline src/StateMachine.ts]: `ErrorBrand` now does nothing. `T | never` is simply `T`. The brand is designed to make string literal types unassignable to errors, which this no longer does.

The branding is needed to prevent some sneaky edge cases. E.g. a user names a state, action or whatever to be an error message: 'Foo has already been declared'.

**Comment by rickwebiii** [comment]
Hmm... the test cases I looked at appear to fail as expected without this change. The only difference I can see is that with your change, the input types don't widen (e.g. `'b'` doesn't become `string`). The error messages are clearer with these changes, but unfortunately the branding is needed to prevent users from doing sneaky things like making a state with the name of an error message, which causes an extends check to pass and breaks the error checking.

Do you have a particular test case I can look at to see the impact of these changes?

File: src/StateMachine.ts [hunk 2]
```diff
@@ -268,9 +270,9 @@ const action = <StateMap, Transitions, ActionMap>(): ActionFunc<StateMap, Transi
 }
 
 const actionHandler = <StateMap, Transitions, ActionMap, HandledStates>(definition: StateMachineDefinition<StateMap, ActionMap>): ActionHandlerFunc<StateMap, Transitions, ActionMap, HandledStates> => {
-  const actionHandlerFunc: ActionHandlerFunc<StateMap, Transitions, ActionMap, HandledStates> = <S extends StateType, AN extends ActionNameType, NS extends MapValues<StateMap>>(
-    state: S,
-    action: AN,
+  return <S extends StateType, AN extends ActionNameType, NS extends MapValues<StateMap>>(
+    state: AssertStateInMap<StateMap, S>,
+    action: AssertActionIsDefined<AN, MapKeys<ActionMap>>,
```
Comment (rickwebiii) [inline src/StateMachine.ts:275]: Ah ok, I think this will give a better error message and I think this might have slipped through the cracks.

File: test/StateMachine.ts [hunk 3]
```diff
@@ -154,42 +154,55 @@ describe('compile-time checking', () => {
     it('should fail when specifying same state twice', () => {
         stateMachine()
             .state<'b'>('b')
-            // @ts-expect-error
-            .state<'b'>('b');
+            .state<'b'>(
+                // @ts-expect-error
```
Comment (rickwebiii) [inline test/StateMachine.ts:158]: 👍 better scoping the assertion

File: test/StateMachine.ts [hunk 4]
```diff
@@ -233,13 +253,16 @@ describe('compile-time checking', () => {
...
-                };
-            });
-            
+            .actionHandler(
+                'a',
+                // @ts-expect-error
+                'a2',
+                () => {
+                    return {
+                        stateName: 'b' as const
```
Comment (rickwebiii) [inline test/StateMachine.ts:262]: Ok, I see now, this is indeed an oversight. This case isn't caught without your change and the reason for the failure is the missing const assertion, not the undefined action.

**Comment by rickwebiii** [comment]
Ok, I see now. `should fail when declaring handler to non-action` wasn't actually catching a genuine error because it was missing an `as const`. Overall, I think if you revert the branding change, everything else can be taken as-is.

Thanks!

**Comment by altano** [comment]
I thought my branding change handled the case where you specified an error message string to circumvent the type checking but it wasn't actually working. Whoops.

I've
- Replaced my `ErrorBrand` change with one that adds unit tests to catch my mistake so no one else makes it in the future
- Changed the code to cast the state and action to get rid of the error about `ErrorBrand` being in the type. This isn't the nicest code but you also had to do it elsewhere in the file already, and it's something we can clean up if throw types ever land in TypeScript.

Let me know if this looks good now, and if not, what else I can do to fix it up to your liking. Thanks again!

**Review by rickwebiii — APPROVED** [review]
LGTM. I'll go ahead and merge this.
